autoblocks/_impl/global_state.py

```diff
@@ -25,10 +29,20 @@ def init() -> None:
         daemon=True,
     )
     background_thread.start()
-
+    signal.signal(signal.SIGTERM, _on_exit)
+    signal.signal(signal.SIGINT, _on_exit)
     _started = True
 
 
+def _on_exit(signum, frame):
+    if _loop and _loop.is_running():
+        log.info(f"Attempting to gracefully stop event loop. Received signal {signum}")
+        _loop.stop()
+        pending = asyncio.all_tasks()
```

Comment:
can we do this only for the send event tasks? how does this work with 1) the main loop itself and 2) while loops like [this](https://github.com/autoblocksai/python-sdk/blob/175730e44294a56aa8a364cb7672c32a43934999/autoblocks/_impl/prompts/manager.py#L143)?

Reply:
For 2 - this code will not work. 

`asyncio.all_tasks` will get all the tasks from the global event loop - including the infinite loop task you created to refresh. 

At first glance I don't see an easy way to tell what the tasks are. 


What do you think about having a dedicated loop for your refresh action? That way I can specifically grab the main event loop which shouldn't contain any tasks that don't resolve

Reply:
i think the issue with that is it will be easy for us to forget the global event loop is off-limits for infinite loops. i think a better way is a dedicated loop for events, but then things are getting pretty complicated. is it possible to attach metadata to futures or otherwise inspect them? they're really just objects and python is the wild west, we might be able to do something to denote they are event futures

Reply:
Mirrored approach here: https://www.roguelynn.com/words/asyncio-graceful-shutdowns/

Was able to use the coros to get tasks that were `_send_event_unsafe`